src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/FlowScannerTest.java

```diff
@@ -0,0 +1,465 @@
+            "sleep 1"
+        ));
+        /** Flow structure (ID - type)
+         2 - FlowStartNode
+         3 - EchoStep
+         4 - TimeoutStep
+         5 - TimeoutStep with BodyInvocationAction
+         6 - EchoStep
+         7 - EchoStep
+         8 - StepEndNode (BlockEndNode), startId=5
+         9 - StepEndNode (BlockEndNode), startId = 4
+         10 - SleepStep
+         11 - FlowEndNode
+         */
+
+        WorkflowRun b = r.assertBuildStatusSuccess(job.scheduleBuild2(0));
+        Predicate<FlowNode> matchEchoStep = FlowTestUtils.predicateMatchStepDescriptor("org.jenkinsci.plugins.workflow.steps.EchoStep");
+        FlowExecution exec = b.getExecution();
+        Collection<FlowNode> heads = exec.getCurrentHeads();
+
+        // Linear analysis
+        LinearScanner linearScanner = new LinearScanner();
+        linearScanner.setup(heads);
+        assertNodeOrder("Linear scan with block", linearScanner, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2);
+        linearScanner.setup(exec.getNode("7"));
+        assertNodeOrder("Linear scan with block from middle ", linearScanner, 7, 6, 5, 4, 3, 2);
+
+        LinearBlockHoppingScanner linearBlockHoppingScanner = new LinearBlockHoppingScanner();
+
+        // // Test block jump core
+        FlowNode headCandidate = exec.getNode("8");
+        Assert.assertEquals(exec.getNode("4"), linearBlockHoppingScanner.jumpBlockScan(headCandidate, Collections.EMPTY_SET));
+        Assert.assertTrue("Setup should return true if we can iterate", linearBlockHoppingScanner.setup(headCandidate, null));
+
+        // Test the actual iteration
+        linearBlockHoppingScanner.setup(heads);
+        Assert.assertFalse(linearBlockHoppingScanner.hasNext());
+        linearBlockHoppingScanner.setup(exec.getNode("8"));
+        assertNodeOrder("Hopping over one block", linearBlockHoppingScanner, 4, 3, 2);
+        linearBlockHoppingScanner.setup(exec.getNode("7"));
+        assertNodeOrder("Hopping over one block", linearBlockHoppingScanner, 7, 6, 5, 4, 3, 2);
+
+        // Test the black list in combination with hopping
+        linearBlockHoppingScanner.setup(exec.getNode("8"), Collections.singleton(exec.getNode("5")));
+        Assert.assertFalse(linearBlockHoppingScanner.hasNext());
+        linearBlockHoppingScanner.setup(exec.getNode("8"), Collections.singleton(exec.getNode("4")));
+        Assert.assertFalse(linearBlockHoppingScanner.hasNext());
+    }
+
+
+    /** And the parallel case */
+    @Test
+    public void testParallelScan() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "Convoluted");
+        job.setDefinition(new CpsFlowDefinition(
+            "echo 'first'\n" +
+            "def steps = [:]\n" +
+            "steps['1'] = {\n" +
+            "    echo 'do 1 stuff'\n" +
+            "}\n" +
+            "steps['2'] = {\n" +
+            "    echo '2a'\n" +
+            "    echo '2b'\n" +
+            "}\n" +
+            "parallel steps\n" +
```

Comment:
BTW when the set of branches is statically known, it is more legible to write (e.g.)

```groovy
parallel one: {
    echo 'do 1 stuff'
}, two: {
    echo '2a'
    echo '2b'
}
```